src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -594,120 +872,69 @@ impl<O: ForestObligation> ObligationForest<O> {
     /// must be run beforehand to remove any cycles on `Success` nodes.
     #[inline(never)]
     fn compress(&mut self, do_completed: DoCompleted) -> Option<Vec<O>> {
-        let orig_nodes_len = self.nodes.len();
-        let mut node_rewrites: Vec<_> = std::mem::take(&mut self.node_rewrites);
-        debug_assert!(node_rewrites.is_empty());
-        node_rewrites.extend(0..orig_nodes_len);
-        let mut dead_nodes = 0;
         let mut removed_done_obligations: Vec<O> = vec![];
 
-        // Move removable nodes to the end, preserving the order of the
-        // remaining nodes.
-        //
-        // LOOP INVARIANT:
-        //     self.nodes[0..index - dead_nodes] are the first remaining nodes
-        //     self.nodes[index - dead_nodes..index] are all dead
-        //     self.nodes[index..] are unchanged
-        for index in 0..orig_nodes_len {
-            let node = &self.nodes[index];
-            match node.state.get() {
-                NodeState::Pending | NodeState::Waiting => {
-                    if dead_nodes > 0 {
-                        self.nodes.swap(index, index - dead_nodes);
-                        node_rewrites[index] -= dead_nodes;
+        // Compress the forest by removing any nodes marked as error or done
+        let mut error_or_done_nodes = mem::take(self.error_or_done_nodes.get_mut());
+        for &index in &error_or_done_nodes {
+            let node = &mut self.nodes[index];
```

Comment:
(Earlier, @nnethercote requested some comments in this code, but it still seems fairly uncommented =) I admit I didn't read it in detail, as it is some fairly dense code. Comments would help. For example, a comment could explain that nodes are being moved into the `dead_nodes` list for later re-use, and maybe talk about the role of `reverse_dependents` and what sort of updates are happening there.